### PR TITLE
Pipeline v4: Formalized Cross Dataset training&testing 

### PR DESCRIPTION
**Pull request by zyzzzz-123**
① you can choose "train_and_test" or "only_test" in yaml. if  "train_and_test", it will train firstly and then use the trained model to test; if "only_test", it will use the pre-trained model to test.
②you can set the parameters of train valid test datasets individually.
③delete evalution_config  some other changes in config to fit the changes of yaml format.
④tscan_trainer and deeppyhs_trainer are updated, but other trainers are kept the same as before. (will be updated in future)

To replicate my result, you can run it with PURE_TSCAN_BASIC.yaml and will get:
best_epoch:  4
loss: 0.6788842607289552
FFT MAE:4.200105773870969
Peak MAE:2.5538677355174695
FFT MAE (Peak Label):4.462451860154304
Peak MAE (Peak Label):1.7392284047725766
FFT MAE (FFT Label):1.1340725806451613
Peak MAE (FFT Label):4.849251368561001

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -14,42 +14,150 @@
...
-_C.DATA.PREPROCESS.H = 128
-_C.DATA.PREPROCESS.DATA_TYPE = ['']
-_C.DATA.PREPROCESS.LABEL_TYPE = ''
+_C.TRAIN.DATA.PREPROCESS = CN()
+_C.TRAIN.DATA.PREPROCESS.DO_CHUNK = True
+_C.TRAIN.DATA.PREPROCESS.CLIP_LENGTH = 180
+_C.TRAIN.DATA.PREPROCESS.DYNAMIC_DETECTION = True
+_C.TRAIN.DATA.PREPROCESS.DETECTION_LENGTH = 180
+_C.TRAIN.DATA.PREPROCESS.CROP_FACE = True
+_C.TRAIN.DATA.PREPROCESS.LARGE_FACE_BOX = True
```
Comment (xliucs) [inline config.py:57]: add a comment to explain what is LARGE_FACE_BOX.

File: config.py [hunk 2]
```diff
@@ -14,42 +14,150 @@
...
+_C.TRAIN.DATA.PREPROCESS.LARGER_BOX_SIZE = 1.5
```
Comment (xliucs) [inline config.py:58]: SAME as BOX SIZE.

File: config.py [hunk 3]
```diff
@@ -14,42 +14,150 @@
...
+_C.TRAIN.DATA.PREPROCESS.W = 128
```
Comment (xliucs) [inline config.py:59]: Actually, let us add a comment to explain every parameter in _C.TRAIN.DATA.PROCESS, but let me help you with these changes.

File: config.py [hunk 4]
```diff
@@ -14,42 +14,150 @@
...
+_C.TEST.DATA.PREPROCESS.CROP_FACE = True
+_C.TEST.DATA.PREPROCESS.LARGE_FACE_BOX = True
+_C.TEST.DATA.PREPROCESS.LARGER_BOX_SIZE = 1.5
+_C.TEST.DATA.PREPROCESS.W = 128
+_C.TEST.DATA.PREPROCESS.H = 128
+_C.TEST.DATA.PREPROCESS.DATA_TYPE = ['']
+_C.TEST.DATA.PREPROCESS.LABEL_TYPE = ''
+
+
+#
```
Comment (xliucs) [inline config.py:126]: If we don't we need these comments, please remove them.

File: config.py [hunk 5]
```diff
@@ -148,78 +233,65 @@ def update_config(config, args):
         config.TRAIN.BATCH_SIZE = args.batch_size
     if args.epochs:
         config.TRAIN.EPOCHS = args.epochs
-    if args.cached_path:
-        config.DATA.CACHED_PATH = args.cached_path
+    # if args.cached_path:
```
Comment (xliucs) [inline config.py:236]: same here.

File: config.py [hunk 6]
```diff
@@ -148,78 +233,65 @@ def update_config(config, args):
...
+    #     config.DATA.CACHED_PATH = args.cached_path
     if args.lr:
         config.TRAIN.LR = args.lr
     if args.model_dir:
         config.MODEL.MODEL_DIR = args.model_dir
-    if args.preprocess:
-        config.DATA.DO_PREPROCESS = args.preprocess
+    # if args.preprocess:
```
Comment (xliucs) [inline config.py:242]: remove

File: dataset/data_loader/BaseLoader.py [hunk 7]
```diff
@@ -160,24 +160,20 @@ def facial_detection(self, frame, larger_box=False, larger_box_size=1.0):
...
-            face_region_all = np.asarray(face_region, dtype='int')
+        if danymic_det:
+            det_num = ceil(frames.shape[0] / det_length)
         else:
-            assert (danymic_det == False)           # danymic_det can be True only when face_detection is True.
-            face_region_all = [0, 0, frames[0].shape[0], frames[0].shape[1]]
+            det_num = 1
+        face_region = list()
 
+        for idx in range(det_num):
```
Comment (xliucs) [inline dataset/data_loader/BaseLoader.py:171]: same here. some comments to explain your for loop would be helpful.

File: dataset/data_loader/BaseLoader.py [hunk 8]
```diff
@@ -232,6 +228,7 @@ def save(self, frames_clips, bvps_clips, filename):
 
     def load(self):
         """Loads the preprocessing data."""
+        print(self.cached_path)
```
Comment (xliucs) [inline dataset/data_loader/BaseLoader.py:231]: if you don't need print this, remove it.

**Review by xliucs — CHANGES_REQUESTED** [review]

File: neural_methods/trainer/TscanTrainer.py [hunk 9]
```diff
@@ -101,27 +106,80 @@ def valid(self, data_loader):
...
+                                                    idx + 1) * config.TEST.DATA.PREPROCESS.CLIP_LENGTH]
+
+        calculate_metrics(predictions, labels, config)
+
+        return 0
+
+
+
+
+    # def test(self, data_loader):
```
[the remark this answers is not in the document]
Comment (zyzzzz-123) [reply neural_methods/trainer/TscanTrainer.py:158]: already deleted.

File: config.py (hunk 5 above)
Comment (zyzzzz-123) [reply config.py:236]: removed.

File: config.py (hunk 4 above)
Comment (zyzzzz-123) [reply config.py:126]: removed.

File: configs/PURE_DEEPPHYS_BASIC.yaml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+NUM_OF_GPU_TRAIN: 4
+LOG:
+  PATH: runs/exp
+MODEL:
+  DROP_RATE: 0.2
+  NAME: DeepPhys
+INFERENCE:
+  BATCH_SIZE: 16
+  MODEL_PATH:  "/data1/acsp/Yuzhe_Zhang/Toolbox_master_2/rPPG-Toolbox/PreTrainedModels/deepphys_synthetics_10epoch_geforce2080ti_Epoch9.pth"
+#  MODEL_PATH:  "/data1/acsp/Yuzhe_Zhang/Toolbox_master_2/rPPG-Toolbox/PreTrainedModels/PURE_SizeW72_SizeH72_ClipLength180_DataTypeNormalized_Standardized_LabelTypeNormalized_Large_boxTrue_Large_size1.5_Dyamic_DetFalse_det_len180/PURE_UBFC_deepphys.pth_Epoch0.pth"
```
[the remark this answers is not in the document]
Comment (zyzzzz-123) [reply configs/PURE_DEEPPHYS_BASIC.yaml:78]: removed.

File: configs/UBFC_TSCAN_BASIC.yaml [hunk 11]
```diff
@@ -1,26 +1,83 @@
...
-  BATCH_SIZE: 4
-  EPOCHS: 30
-  LR: 0.0001
-  MODEL_FILE_NAME: tscan.pth
-  
\ No newline at end of file
+INFERENCE:
+  BATCH_SIZE: 16
+  MODEL_PATH:   ""
+        #MODEL_PATH: # if TRAIN_OR_TEST:"train_and_test", should be "" or anything, it will use currently trained model.
```
[the remark this answers is not in the document]
Comment (zyzzzz-123) [reply configs/UBFC_TSCAN_BASIC.yaml:80]: removed.

File: dataset/data_loader/BaseLoader.py [hunk 12]
```diff
@@ -38,7 +38,8 @@ def __init__(self, name, data_path, config_data):
         """
         self.name = name
         self.data_path = data_path
-        self.cached_path = os.path.join(config_data.CACHED_PATH, name)
+        # self.cached_path = os.path.join(config_data.CACHED_PATH, name)
```
[the remark this answers is not in the document]
Comment (zyzzzz-123) [reply dataset/data_loader/BaseLoader.py:41]: removed.

File: dataset/data_loader/BaseLoader.py [hunk 13]
```diff
@@ -160,24 +160,20 @@ def facial_detection(self, frame, larger_box=False, larger_box_size=1.0):
         return result
 
     def resize(self, frames, danymic_det, det_length,
```
[the remark this answers is not in the document]
Comment (zyzzzz-123) [reply dataset/data_loader/BaseLoader.py:162]: added. Can you review these comments, please? I' 'm worried that they are not clear enough.

File: dataset/data_loader/BaseLoader.py (hunk 7 above)
Comment (zyzzzz-123) [reply dataset/data_loader/BaseLoader.py:171]: I have added some comments. Please review them.

File: dataset/data_loader/BaseLoader.py (hunk 8 above)
Comment (zyzzzz-123) [reply dataset/data_loader/BaseLoader.py:231]: removed.

File: neural_methods/trainer/DeepPhysTrainer.py [hunk 14]
```diff
@@ -106,19 +106,27 @@ def test(self, data_loader):
...
-                state_dict = checkpoint
-                new_state_dict = OrderedDict()
-                for k, v in state_dict.items():
-                    name = k[7:]  # remove 'module.' of dataparallel
-                    new_state_dict[name] = v
-                self.model.load_state_dict(new_state_dict)
-            else:
-                self.model.load_state_dict(torch.load(config.INFERENCE.MODEL_PATH))
-            self.model = self.model.to(config.DEVICE)
+            # if config.NUM_OF_GPU_TRAIN > 1:
```
[the remark this answers is not in the document]
Comment (zyzzzz-123) [reply neural_methods/trainer/DeepPhysTrainer.py:109]: removed.

**Comment by zyzzzz-123** [comment]
The latest commit fixes the dataset load error detection problem, it will assert an error if the data_path is not true.

**Comment by zyzzzz-123** [comment]
①I formalize yamls about  deeppyhs and tscan and do some testing experiment using these yanls.
You can refer to https://www.notion.so/pipeline_v4-experiment-document-02463b14111e44e1b0518ebbce07ac19 to see specific results and replicate them.
② delete some useless files. Adds and deletes some comments.
③ there seems to be no need to maintain evaluation_neural_method.py, delete it or not?
Some problems:
1.deepphys training results change when num_of_GPU changes, it seems to have a bug.
2.tscan doesn’t support multi-gpu right now.
3.It seems that when testing on UBFC, although using **different trained models** and **different preprocessing methods**(previous experiments, not here), the results are partially the same.
I am still working on visualizing it to figure out whether some bugs account for it or it’s just a unique attribute of UBFC.

**Review by xliucs — APPROVED** [review]
LGTM
